### PR TITLE
pythonPackages.mesa: init at 0.8.6

### PR DESCRIPTION
**Pull request by dpaetzel**
###### Motivation for this change

mesa is a Python 3+ library for creating multi-agent simulations.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -1736,7 +1736,7 @@
     name = "Chris Double";
   };
   dpaetzel = {
-    email = "david.a.paetzel@gmail.com";
+    email = "david.paetzel@posteo.de";
```
Comment (jonringer) [inline maintainers/maintainer-list.nix]: this should be a separate commit, and probably a different PR

File: pkgs/development/python-modules/mesa/default.nix [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    owner = "projectmesa";
+    rev = "v${version}";
+    sha256 = "0d8c636zhswxd91ldlmdxxlyym2fj3bk1iqmpc1jp3hg7vvc7w03";
+  };
+
+  # According to their docs, this library is for Python 3+.
+  disabled = isPy27;
+
+  propagatedBuildInputs = [ cookiecutter networkx pandas tornado tqdm ];
+
```
Comment (jonringer) [inline pkgs/development/python-modules/mesa/default.nix]: Please add tests
```suggestion
checkInputs = [ pytest ];
checkPhase = ''
  pytest tests
'';
```

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (dpaetzel) [reply maintainers/maintainer-list.nix]: OK, different PR then.

File: pkgs/development/python-modules/mesa/default.nix (hunk 2 above)
Comment (dpaetzel) [reply pkgs/development/python-modules/mesa/default.nix]: Yes, of course. I just did, but as of now, one of the tests fails. I think it has something to do with it being run by Nix because it tests whether a directory is correctly created by some procedure; for that it uses 'click.testing'’s `runner.isolated_filesystem` (which is used to run code in a new temporary folder). However, the test does not find the created directory afterwards. Perhaps you can give me a hint?

Comment (jonringer) [reply pkgs/development/python-modules/mesa/default.nix]: do you mind pasting the results?

Comment (jonringer) [reply pkgs/development/python-modules/mesa/default.nix]: actually, i did it for you:

```
checkInputs = [ pytest ];

  checkPhase = ''
    pytest tests -k 'not scaffold'
  '';
```

with pytest, you can use the `-k` option to specify a predicate on test case names, or you can use `--ignore=path/to/test/file.py` to ignore entire files/directories.

Comment (dpaetzel) [reply pkgs/development/python-modules/mesa/default.nix]: Ah, OK, thanks a lot. I did not know that simply disabling the problematic test was a viable option.

Comment (jonringer) [reply pkgs/development/python-modules/mesa/default.nix]: it is if it's something out of our control. Since nixpkgs builds in a sandbox environment, a lot of system changing things may fail. Which is fine. 

At the very least, we should be trying to enable as many tests as we can.

File: pkgs/development/python-modules/mesa/default.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  src = fetchFromGitHub {
+    repo = "mesa";
+    owner = "projectmesa";
+    rev = "v${version}";
+    sha256 = "0d8c636zhswxd91ldlmdxxlyym2fj3bk1iqmpc1jp3hg7vvc7w03";
+  };
+
+  # According to their docs, this library is for Python 3+.
+  disabled = isPy27;
```
Comment (jonringer) [inline pkgs/development/python-modules/mesa/default.nix]: i would put disabled close below version

File: pkgs/development/python-modules/mesa/default.nix [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+  checkInputs = [ pytest ];
+
+  checkPhase = ''
```
Comment (jonringer) [inline pkgs/development/python-modules/mesa/default.nix]: we should comment as to why this is disabled
```suggestion
  # ignore test which tries to mkdir in unreachable location
  checkPhase = ''
```

File: pkgs/development/python-modules/mesa/default.nix (hunk 3 above)
Comment (dpaetzel) [reply pkgs/development/python-modules/mesa/default.nix]: Sounds good.

File: pkgs/development/python-modules/mesa/default.nix (hunk 4 above)
Comment (dpaetzel) [reply pkgs/development/python-modules/mesa/default.nix]: Yes, definitely.

File: pkgs/development/python-modules/mesa/default.nix [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    repo = "mesa";
+    owner = "projectmesa";
+    rev = "v${version}";
+    sha256 = "0d8c636zhswxd91ldlmdxxlyym2fj3bk1iqmpc1jp3hg7vvc7w03";
+  };
+
+  checkInputs = [ pytest ];
+
+  # Ignore test which tries to mkdir in unreachable location.
+  checkPhase = ''
```
Comment (flokli) [inline pkgs/development/python-modules/mesa/default.nix]: you can skip the custom test phase, and set `disabledTests = [ "scaffold" ];` instead.

**Review by flokli — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/mesa/default.nix [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  disabled = isPy27;
+
+  src = fetchFromGitHub {
+    repo = "mesa";
+    owner = "projectmesa";
+    rev = "v${version}";
+    sha256 = "0d8c636zhswxd91ldlmdxxlyym2fj3bk1iqmpc1jp3hg7vvc7w03";
+  };
+
+  checkInputs = [ pytest ];
```
[the remark this answers is not in the document]
Comment (dpaetzel) [reply pkgs/development/python-modules/mesa/default.nix]: OK, I like this. However, this way, `pytest` also runs tests for the examples provided with the library which don't run through due to the following error:
```
_____ ERROR collecting examples/wolf_sheep/wolf_sheep/test_random_walk.py ______
ImportError while importing test module '/build/source/examples/wolf_sheep/wolf_sheep/test_random_walk.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
examples/wolf_sheep/wolf_sheep/test_random_walk.py:11: in <module>
    from random_walk import RandomWalker
E   ModuleNotFoundError: No module named 'random_walk'
```

There seems to be an issue with the path? Any suggestions?

[the remark this answers is not in the document]
Comment (dpaetzel) [reply pkgs/development/python-modules/mesa/default.nix]: This might be a case of the “`pytest` vs. `python -m pytest`” consideration?

**Comment by dpaetzel** [comment]
I disabled the tests for the `examples` directory (but it feels a bit to hacky, to be honest).

File: pkgs/development/python-modules/mesa/default.nix [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  # Disable tests of the examples because of import errors (due to directory
+  # layout?).
+  preCheck = ''
+    cat > conftest.py <<EOF
+    import sys
+    collect_ignore_glob = ["examples/*"]
+    EOF
+  '';
+
+  checkInputs = [ pytest pytestCheckHook ];
```
Comment (FRidh) [inline pkgs/development/python-modules/mesa/default.nix]: whn adding the hook we also need to pass `dontUseSetuptoolsCheck = true;` as it will otherwise try to run two phases.

File: pkgs/development/python-modules/mesa/default.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (flokli) [reply pkgs/development/python-modules/mesa/default.nix]: @FRidh, wdyt?

[the remark this answers is not in the document]
Comment (jonringer) [reply pkgs/development/python-modules/mesa/default.nix]: you can see that the fil is being collected from the examples directory. you want:
```
checkPhase = ''
  pytest tests
'';
```
so that you only bring in examples from the tests directory

[the remark this answers is not in the document]
Comment (dpaetzel) [reply pkgs/development/python-modules/mesa/default.nix]: And, correspondingly, I don't use `pytestCheckHook`?

[the remark this answers is not in the document]
Comment (jonringer) [reply pkgs/development/python-modules/mesa/default.nix]: correct to both

**Comment by flokli** [comment]
@GrahamcOfBorg build python3.pkgs.mesa

File: pkgs/development/python-modules/mesa/default.nix [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+buildPythonPackage rec {
+  pname = "mesa";
+  version = "0.8.6";
+
+  # According to their docs, this library is for Python 3+.
+  disabled = isPy27;
+
+  src = fetchFromGitHub {
+    repo = "mesa";
```
Comment (flokli) [inline pkgs/development/python-modules/mesa/default.nix]: can you flip owner and repo? Usually, it's the other way round :-D

Comment (dpaetzel) [reply pkgs/development/python-modules/mesa/default.nix]: Oh, sure. Don't know how that happened :laughing:

**Comment by jonringer** [comment]
do you mind rebasing this on top of master... I seem to be building things that I shouldn't when I review...

```
git pull -r origin master
git push --force .. ..
```

**Comment by dpaetzel** [comment]
Yes, sure. :slightly_smiling_face:

**Comment by jonringer** [comment]
figured it out, pyside is importing the mesa package, before it was getting it from the pkgs package scope, but now it's finding in python packages.

```
$ rg "mesa "
pyside/default.nix
1:{ lib, fetchurl, cmake, buildPythonPackage, pysideGeneratorrunner, pysideShiboken, qt4, mesa }:
18:  buildInputs = [ mesa ];
```

do you mind adding doing:
```
  pyside = callPackage ../development/python-modules/pyside {
    inherit (pkgs) mesa;
  };
```
to top-level/python-modules.nix and naming it: `pythonPackages.pyside: inherit mesa from pkgs`

**Comment by risicle** [comment]
`nix-review` is happy for me, macos 10.13.

**Comment by flokli** [comment]
@dpaetzel can you address the changes requested from @jonringer? Looks good for me to merge then :-)

**Comment by dpaetzel** [comment]
I finally got to address the `pyside` request. Sorry it took so long.

**Review by jonringer — APPROVED** [review]
`nix-review` passes on NixOS (python38Packages.pytest is broken on master)
diff LGTM
outputs LGTM
able to import from python interpreter

```
[2 built (4 failed), 8 copied (0.8 MiB), 0.7 MiB DL]
error: build of '/nix/store/2nni8zqp3cgyn0k12h856j7nalgl9f2w-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/71244
1 package failed to build:
python38Packages.mesa

1 package were build:
python37Packages.mesa
```
```
[nix-shell:/home/jon/.cache/nix-review/pr-71244-3]$ nix path-info -Sh ./results/python37Packages.mesa
/nix/store/rs9hnidl4zrakl1ka6kwbvdfgdaaxg2c-python3.7-mesa-0.8.6         403.0M
# mostly due to pandas closure
[nix-shell:/home/jon/.cache/nix-review/pr-71244-3]$ nix path-info -Sh ./result
/nix/store/aml9daqjz8c7z33k3j2sg12p622z50hd-python3.7-pandas-0.25.1      371.6M
```

**Comment by jonringer** [comment]
@GrahamcOfBorg build python3Packages.mesa

**Comment by dpaetzel** [comment]
Oh, I probably should exclude `mesa` from `python38Packages` then?

**Comment by jonringer** [comment]
No, python38Packages.pytest is broken on master, so I'm disregarding it for now

**Comment by jonringer** [comment]
it was fixed in staging, just hasn't made it's way to master yet

**Comment by jonringer** [comment]
I was wondering why the closure was so huge, i managed to shave off 57MB #71553

**Comment by dpaetzel** [comment]
Ah, that clarifies it. Nice work on the closure size!

**Comment by jonringer** [comment]
ofborg finally built everything :), merging.

Thanks for sticking through all the pain

**Comment by dpaetzel** [comment]
Nice! Thank you, reviewers :slightly_smiling_face:
